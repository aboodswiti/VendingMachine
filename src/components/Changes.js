import React from "react";

const Changes = ({ allMoney, availableChange, returnedChange }) => {
  // Display Total money & availableChange inside VM
  return (
    <div className="changes">
      <div>
        {Object.keys(returnedChange).length ? <p>Change : </p> : null}
        {Object.entries(returnedChange).map(([key, val]) => (
          <p key={key}>
            ${key}: {val}
          </p>
        ))}
      </div>

      <p>
        Available
        <br />
        Change :
      </p>
      <h2>${allMoney}</h2>

      <h4> $Type : #Amount </h4>
      {Object.entries(availableChange).map(([key, val]) => (
        <h3 key={key}>
          ${key}: {val}
        </h3>
      ))}
    </div>
  );
};

export default Changes;
