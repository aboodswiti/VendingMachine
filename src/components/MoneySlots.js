import React from "react";
import { Button } from "@material-ui/core";

const MoneySlots = ({ addMoney }) => {
  let notes = [20, 50];
  let coins = [0.1, 0.2, 0.5, 1];
  
 // display the coin and note slots 
  return (
    <div className="moneySlots">
      <div className="CoinSlot">
        <h2> Coins Slots : </h2>

        {coins.map((coin, i) => (
          <Button
            key={i}
            data-testid={`coin-${coin}`}
            style={{
              height: "3.5rem",
              margin: "9px",
              borderRadius: "60%",
              fontSize: "1.1rem",
            }}
            variant="contained"
            color="secondary"
            onClick={() => addMoney(coin)}
          >
            {coin < 1 ? `${coin * 100}Â¢` : `$${coin}`}
          </Button>
        ))}
      </div>
      <div className="NotesSlot ">
        <h2> Notes Slot : </h2>

        {notes.map((note, i) => (
          <Button
            key={i}
            style={{
              margin: "9px",
              fontSize: "1.5rem",
            }}
            variant="outlined"
            color="secondary"
            onClick={() => addMoney(note)}
          >
            {`$${note}`}
          </Button>
        ))}
      </div>
    </div>
  );
};

export default MoneySlots;
