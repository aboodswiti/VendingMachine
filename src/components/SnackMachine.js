import React, { useState, useEffect } from "react";
import { snacks } from "../snacks";
import SnackSlots from "./SnackSlots";
import MoneySlots from "./MoneySlots";
import Keypad from "./Keypad";
import { Button } from "@material-ui/core";
import ChangeMaking from "./ChangeMaking";
import Changes from "./Changes";
const SnackMachine = () => {
  const [location, setLocation] = useState("");
  const [isItemSelected, setIsItemSelected] = useState(false);
  const [res, setRes] = useState(Array);
  const [amount, setAmount] = useState(0);
  const [isEnough, setIsEnough] = useState(false);
  const [change, setChange] = useState(Number);
  const [returnedChange, setReturnedChange] = useState(Number);
  const [isThereChange, setIsThereChange] = useState(false);
  const [allMoney, setAllMoney] = useState(0);
  const [items, setItems] = useState([...snacks]);
  const [availableChange, setavailableChange] = useState({
    1: 17,
    20: 4,
    50: 3,
    0.1: 8,
    0.2: 5,
    0.5: 4,
  });

  useEffect(() => {
    let result = items.filter((item) => item.location === location); // get selected snack
    let sum = 0;
    setItems(    // If(quantity = 0) the item is sold out , display Sold out img of this item  
      items.map((item) =>
        item.quantity === 0
          ? {
              ...item,
              img: "https://toppng.com/uploads/preview/sold-out-stamp-11523435603ivg6vbqdjd.png",
            }
          : item
      )
    );
    if (  // check it the item not exist 
      result.length === 0 ||
      result[0] === "undefined" ||
      result[0].quantity === 0
    )
      setIsItemSelected(false);
    else setIsItemSelected(true);
    if (isItemSelected && amount >= res.price) { // Validate item price and insert money & calculate the change
      setIsEnough(true);
      let change = amount - res.price;
      setChange(parseFloat(change.toFixed(2), 10));
    } else setIsEnough(false);

    setRes(result[0]);
    Object.entries(availableChange).map( // calculate total money inside Vending machine
      ([key, val]) => (sum = sum + key * val)
    );
    setAllMoney(parseFloat(sum.toFixed(2), 10));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [res, amount, location, isItemSelected, isEnough, availableChange]);

  const addMoney = (value) => { // to add money to vending mashine when customer insert money
    let money = amount + value;
    setAmount(parseFloat(money.toFixed(2), 10)); 

    setavailableChange((prevState) => ({ // update the availableChange with money type inside VM when insert money
      ...prevState,
      [value]: availableChange[value] + 1,
    }));
  };

  const purchaseItem = (location, quantity) => { // purchase the item and update the quantity of this item
    setAmount(change);
    setRes((res.length = 0));
    setLocation("");
    setItems(
      items.map((item) =>
        item.location === location ? { ...item, quantity: quantity - 1 } : item
      )
    );
  };

  return (
    <div data-testid="VM-container" className="container">
      {items.map((item, i) => (  // display all snacks 
        <div className="item" key={i}>
          <SnackSlots key={i} item={item} />
        </div>
      ))}
      <div className="vm_container">
        <div className="Selected-item">  
          {isItemSelected ? ( // display the selected snack
            <SnackSlots key={location} item={res} />
          ) : location !== "" ? (
            <p> Does not exist! </p> //  display this message if the selected item does not exist
          ) : null}
          <h2 data-testid="AvailableMoney">Available Money : ${amount}</h2>  

          {isEnough ? ( // Hide Purchase button if the money is not Enough
            <div>
              <Button
                disabled={!isThereChange}
                style={{ margin: "1rem" }}
                variant="outlined"
                color="secondary"
                onClick={() => purchaseItem(res?.location, res?.quantity)}
              >
                Purchase the {res?.item}
              </Button>
            </div>
          ) : null}

          <ChangeMaking
            availableChange={availableChange}
            change={change}
            isItemSelected={isItemSelected}
            setavailableChange={setavailableChange}
            setAmount={setAmount}
            setIsThereChange={setIsThereChange}
            isThereChange={isThereChange}
            amount={amount}
            location={location}
            setReturnedChange={setReturnedChange}
            isEnough={isEnough}
            setLocation={setLocation}
          />
        </div>

        <Keypad
          location={location}
          setLocation={setLocation}
          isItemSelected={isItemSelected}
          setReturnedChange={setReturnedChange}
        />
        <MoneySlots addMoney={addMoney} />
      </div>
      <Changes
        allMoney={allMoney}
        availableChange={availableChange}
        returnedChange={returnedChange}
      />
    </div>
  );
};

export default SnackMachine;
