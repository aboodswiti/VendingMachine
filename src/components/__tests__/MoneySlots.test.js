import React from 'react';
import ReactDOM from 'react-dom';
import {render, fireEvent, cleanup,screen} from '@testing-library/react';
import renderer from "react-test-renderer"
import MoneySlots from "../MoneySlots"
import SnackMachine from "../SnackMachine"
import Changes from "../Changes"
afterEach(() => {
    cleanup();
});


test('Should render coin slots', ()=> {
    render(<MoneySlots/>)
    const CoinElement = screen.getByText("Coins Slots :")
expect(CoinElement).toBeInTheDocument()

})
test('Should render Notes slots', ()=> {
    render(<MoneySlots/>)
    const NotesElement = screen.getByText("Notes Slot :")
expect(NotesElement).toBeInTheDocument()

})

test('Should call addmoney', ()=> {
    const addMoney = jest.fn()
    render(<MoneySlots addMoney={addMoney} />)

    const Add20CentsButton = screen.getByTestId("coin-0.2")
    fireEvent.click(Add20CentsButton)

    expect(addMoney).toHaveBeenCalledTimes(1)
    fireEvent.click(Add20CentsButton)
    expect(addMoney).toHaveBeenCalledTimes(2)


})

test('MoneySlots, Should Matches snapshot', ()=> {
const addMoney = jest.fn()

const tree = renderer.create(<MoneySlots addMoney={addMoney} />).toJSON();
expect(tree).toMatchSnapshot();
} );



it('Available Money in state is $0 at initial state', () => {
    const { getByText } = render(<SnackMachine />);
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $0")


 })

 it('Add money by 20¢ should changes the Available Money, amount state to be $0.2', () => {
    const { getByText } = render(<SnackMachine>
                                  <MoneySlots />
                                 </SnackMachine>)
  
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $0")

    fireEvent.click(getByText("20¢"))
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $0.2")

  })
  

  it('Add money by 20¢ twice should changes the Available Money, amount state to be $0.4', () => {
    const { getByText } = render(<SnackMachine>
                                  <MoneySlots />
                                 </SnackMachine>)
  
    //expect(getByText(/Moe/i).textContent).toBe("Moe")
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $0")

    fireEvent.click(getByText("20¢"))
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $0.2")
    fireEvent.click(getByText("20¢"))
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $0.4")

    // expect(getByText(/Steve/i).textContent).toBe("Steve")
  })
  
  it('Add money by $20  should changes the Available Money, amount state to be $20', () => {
    const { getByText } = render(<SnackMachine>
                                  <MoneySlots />
                                 </SnackMachine>)
  
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $0")

    fireEvent.click(getByText("$20"))
    expect(getByText(/Available Money /i).textContent).toBe("Available Money : $20")

  })

  it('Add money by $20  should changes the AllMoney to be $270.8', () => {
    const { getByText } = render(<SnackMachine>
                                  <MoneySlots />
                                  <Changes/>
                                 </SnackMachine>)
  
    expect(getByText("$250.8").textContent).toBe("$250.8")

    fireEvent.click(getByText("$20"))
    expect(getByText("$270.8").textContent).toBe("$270.8")

  })

