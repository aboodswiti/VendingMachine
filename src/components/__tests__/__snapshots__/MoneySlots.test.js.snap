// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoneySlots, Should Matches snapshot 1`] = `
<div
  className="moneySlots"
>
  <div
    className="CoinSlot"
  >
    <h2>
       Coins Slots : 
    </h2>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      data-testid="coin-0.1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "borderRadius": "60%",
          "fontSize": "1.1rem",
          "height": "3.5rem",
          "margin": "9px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        10¢
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      data-testid="coin-0.2"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "borderRadius": "60%",
          "fontSize": "1.1rem",
          "height": "3.5rem",
          "margin": "9px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        20¢
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      data-testid="coin-0.5"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "borderRadius": "60%",
          "fontSize": "1.1rem",
          "height": "3.5rem",
          "margin": "9px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        50¢
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      data-testid="coin-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "borderRadius": "60%",
          "fontSize": "1.1rem",
          "height": "3.5rem",
          "margin": "9px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        $1
      </span>
    </button>
  </div>
  <div
    className="NotesSlot "
  >
    <h2>
       Notes Slot : 
    </h2>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "fontSize": "1.5rem",
          "margin": "9px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        $20
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "fontSize": "1.5rem",
          "margin": "9px",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        $50
      </span>
    </button>
  </div>
</div>
`;
